File: Coding-Interviews/30-包含min函数的栈.cpp
#include <stack>
using namespace std;
class MinStack
{
public:
    /** initialize your data structure here. */
    MinStack() {}

    void push(int x)
    {
        count++;
        data.push(x);
        if (min_cur.empty())
            min_cur.push(x);
        else
            min_cur.push(::min(x, min_cur.top()));
    }

    void pop()
    {
        if (!count)
            return;
        count--;
        min_cur.pop();
        data.pop();
    }

    int top() { return data.top(); }

    int min() { return min_cur.top(); }

private:
    int        count{0};
    stack<int> data;
    stack<int> min_cur;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->min();
 */

File: Coding-Interviews/30-包含min函数的栈.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        # data 和 minData 是一一对应的关系
        self.data = []
        self.minData = []

    def push(self, x: int) -> None:
        if not self.data or x < self.min():
            self.minData.append(x)
        else:
            self.minData.append(self.minData(-1))
        self.data.append(x)

    def pop(self) -> None:
        self.data.pop()
        self.minData.pop()

    def top(self) -> int:
        return self.data[-1]

    def min(self) -> int:
        return self.minData[-1]

        # Your MinStack object will be instantiated and called as such:
        # obj = MinStack()
        # obj.push(x)
        # obj.pop()
        # param_3 = obj.top()
        # param_4 = obj.min()
